File: app/model/forecast.py
"""风险预测模型（Phase 5 智能化预测 M1）。

对 ``RiskHealthSnapshot`` 的时序（如项目 ``risk_index`` 日序列）做趋势外推，
把「未来 N 天预测值」结构化落库，供预测列表、预测性预警（M3）与驾驶舱
预测卡（M4）复用。

每个 (scope_type, ref_id, metric, horizon_days) 只保留一条最新预测（upsert），
历史预测不留存 —— 需要回看时可由快照序列随时重算。
数据范围：经 project_id 走 VIA_PROJECT（在 app.core.data_scope 注册）。
"""

from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Index, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.model.base import Base, TimestampMixin
from app.model.project import Project


class Forecast(Base, TimestampMixin):
    __tablename__ = "forecast"

    # 归属项目（scope_type=device 时也回填其归属项目，供数据范围过滤）
    project_id: Mapped[int | None] = mapped_column(
        ForeignKey("project.id", ondelete="SET NULL"), nullable=True, index=True
    )
    project: Mapped["Project"] = relationship("Project", lazy="selectin")

    scope_type: Mapped[str] = mapped_column(String(16), nullable=False, comment="project|device")
    ref_id: Mapped[str] = mapped_column(
        String(64), nullable=False, comment="project.id 或 device_no"
    )
    name: Mapped[str | None] = mapped_column(
        String(128), nullable=True, comment="项目名称 / 设备编号"
    )
    metric: Mapped[str] = mapped_column(
        String(32), nullable=False, default="risk_index", comment="预测指标(risk_index 等)"
    )

    horizon_days: Mapped[int] = mapped_column(Integer, nullable=False, comment="预测跨度(天)")
    sample_count: Mapped[int] = mapped_column(Integer, nullable=False, comment="拟合样本数")

    last_value: Mapped[float] = mapped_column(Float, nullable=False, comment="序列最新观测值")
    slope: Mapped[float] = mapped_column(Float, nullable=False, comment="OLS 斜率(每天变化量)")
    intercept: Mapped[float] = mapped_column(Float, nullable=False, comment="OLS 截距")
    forecast_value: Mapped[float] = mapped_column(
        Float, nullable=False, comment="外推预测值(截断到 0-100)"
    )
    forecast_level: Mapped[str | None] = mapped_column(
        String(8), nullable=True, comment="预测级别(risk:高/中/低; health:优/良/中/差)"
    )

    # 置信带（M2）：残差标准差 + 95% 置信上下界（均截断 0-100）
    std_resid: Mapped[float | None] = mapped_column(
        Float, nullable=True, comment="OLS 残差标准差(样本≤2 时为 0)"
    )
    forecast_lower: Mapped[float | None] = mapped_column(
        Float, nullable=True, comment="95% 置信下界"
    )
    forecast_upper: Mapped[float | None] = mapped_column(
        Float, nullable=True, comment="95% 置信上界"
    )

    forecast_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, comment="预测目标时刻(最后观测+horizon)"
    )
    computed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, comment="本次计算时刻"
    )

    __table_args__ = (
        Index("ix_forecast_project", "project_id"),
        Index("ix_forecast_key", "scope_type", "ref_id", "metric", "horizon_days", unique=True),
    )

File: web/src/api/forecast.ts
// 风险预测（Phase 5 智能化预测）前端接口
// 与后端 app/api/v1/forecasts.py 对齐：
// - GET  /v1/forecasts          预测列表（scope_type/metric/project_id 过滤）
// - GET  /v1/forecasts/preview  单对象序列预览（历史点+拟合+预测点+置信带）
// - POST /v1/forecasts/recompute 重算
import { http } from "@/utils/request";

// forecast 表一行（列表项）
export interface ForecastItem {
  id: number;
  project_id: number | null;
  scope_type: "project" | "device";
  ref_id: string;
  name: string | null;
  metric: "risk_index" | "health_score";
  horizon_days: number;
  sample_count: number;
  last_value: number;
  slope: number;
  intercept: number;
  forecast_value: number;
  forecast_level: string | null; // risk: 高/中/低；health: 优/良/中/差
  std_resid: number | null;
  forecast_lower: number | null;
  forecast_upper: number | null;
  forecast_at: string;
  computed_at: string;
}

// preview 的历史序列点
export interface ForecastSeriesPoint {
  at: string;
  value: number;
}

// preview 的拟合结果（样本不足时为 null）
export interface ForecastFit {
  metric: string;
  horizon_days: number;
  sample_count: number;
  last_value: number;
  slope: number;
  intercept: number;
  forecast_value: number;
  forecast_level: string | null;
  std_resid: number;
  forecast_lower: number;
  forecast_upper: number;
  forecast_at: string;
  computed_at: string;
}

export interface ForecastPreview {
  scope_type: string;
  ref_id: string;
  metric: string;
  horizon_days: number;
  series: ForecastSeriesPoint[];
  forecast: ForecastFit | null;
}

export interface ListForecastParams {
  project_id?: number;
  scope_type?: "project" | "device";
  metric?: "risk_index" | "health_score";
}

export function listForecasts(params: ListForecastParams = {}): Promise<{ items: ForecastItem[] }> {
  return http<{ items: ForecastItem[] }>({ url: "/v1/forecasts/", method: "GET", params });
}

export function previewForecast(params: {
  ref_id: string;
  scope_type: "project" | "device";
  horizon_days?: number;
  history_days?: number;
}): Promise<ForecastPreview> {
  return http<ForecastPreview>({ url: "/v1/forecasts/preview", method: "GET", params });
}

export function recomputeForecasts(params: { project_id?: number; horizon_days?: number } = {}): Promise<
  ForecastItem | { computed: number; skipped: number; projects: number; devices: number }
> {
  return http({ url: "/v1/forecasts/recompute", method: "POST", params });
}

// 预测命中率报表结果（预测性预警闭环验证）
export interface PredictionHitRate {
  verifiable: number;
  hits: number;
  false_positives: number;
  pending: number;
  hit_rate: number | null;
  avg_lead_hours: number | null;
  by_metric: Record<string, { metric: string; verifiable: number; hits: number; false_positives: number; pending: number; hit_rate: number | null }>;
  by_project: { project_id: number; verifiable: number; hits: number; false_positives: number; pending: number; hit_rate: number | null }[];
  period_days: number;
  generated_at: string;
}

export function getForecastHitRate(params: {
  days?: number;
  project_id?: number;
  metric?: "risk_index" | "health_score";
} = {}): Promise<PredictionHitRate> {
  return http<PredictionHitRate>({ url: "/v1/forecasts/hit-rate", method: "GET", params });
}

// 回测摘要（POST /v1/forecasts/backtest 返回）
export interface BacktestSummary {
  models: string[];
  anchors: number;
  rows: number;
  by_model: Record<string, { rows: number; hits: number; false_positives: number }>;
  horizon_days: number;
}

export function runForecastBacktest(params: {
  days?: number;
  horizon_days?: number;
} = {}): Promise<BacktestSummary> {
  return http<BacktestSummary>({ url: "/v1/forecasts/backtest", method: "POST", params });
}

// A/B 命中率报表（GET /v1/forecasts/hit-rate/ab 返回）
export interface ABModelRow {
  model_version: string;
  label: string;
  verifiable: number;
  hits: number;
  false_positives: number;
  pending: number;
  hit_rate: number | null;
  false_positive_rate: number | null;
  avg_lead_hours: number | null;
  by_metric: Record<
    string,
    {
      metric: string;
      verifiable: number;
      hits: number;
      false_positives: number;
      pending: number;
      hit_rate: number | null;
    }
  >;
}

export interface ABComparison {
  baseline: string;
  baseline_label: string;
  challenger: string;
  challenger_label: string;
  hit_rate_baseline: number | null;
  hit_rate_challenger: number | null;
  hit_rate_delta: number | null;
  hit_rate_delta_pct: number | null;
  false_positive_rate_baseline: number | null;
  false_positive_rate_challenger: number | null;
  false_positive_rate_delta: number | null;
  avg_lead_hours_baseline: number | null;
  avg_lead_hours_challenger: number | null;
  lead_delta_hours: number | null;
  better: boolean;
  summary: string;
}

export interface ABHitRate {
  period_days: number;
  generated_at: string;
  models: ABModelRow[];
  comparison: ABComparison | null;
}

export function getForecastABHitRate(params: {
  days?: number;
  project_id?: number;
  metric?: "risk_index" | "health_score";
} = {}): Promise<ABHitRate> {
  return http<ABHitRate>({ url: "/v1/forecasts/hit-rate/ab", method: "GET", params });
}

// 当前上线默认预测模型（GET /v1/forecasts/model/default 返回）
export interface ForecastDefaultModel {
  model_version: string;
  available: { model_version: string; label: string }[];
}

export function getForecastDefaultModel(): Promise<ForecastDefaultModel> {
  return http<ForecastDefaultModel>({ url: "/v1/forecasts/model/default", method: "GET" });
}

export function setForecastDefaultModel(model_version: string): Promise<{
  model_version: string;
  label: string;
  recomputed: unknown;
  predictive_alerts: unknown;
}> {
  return http({
    url: "/v1/forecasts/model/default",
    method: "POST",
    data: { model_version },
  });
}
